### PR TITLE
refactor: fix clippy warnings, replace Arc<Mutex> with Rc<RefCell> in CoreAudio

### PR DESCRIPTION
**Pull request by roderickvd**
Fixex standard clippy warnings, as well as a more impactful change in CoreAudio (macOS): to resolve the `clippy::arc-with-non-send-sync` warning, it replaces `Arc<Mutex<StreamInner>>` with `Rc<RefCell<StreamInner>>`.

## CoreAudio changes (review requested)
The main change replaces `Arc<Mutex<StreamInner>>` with `Rc<RefCell<StreamInner>>` in the macOS CoreAudio implementation. Rationale:

- `StreamInner` contains `AudioUnit` and `AudioObjectPropertyListener` which are not `Send`/`Sync`
- The `Arc<Mutex<>>` provided no thread-safety benefits since the inner types aren't thread-safe
- Stream usage patterns in CPAL are single-threaded
- `Rc<RefCell<>>` is more appropriate for single-threaded shared ownership

This changes fundamental synchronization primitives in the CoreAudio backend. While the change is theoretically sound and all tests pass, extra scrutiny would be appreciated to ensure that we are not missing some edge cases.

### CODE REVIEW

**Comment by roderickvd** [comment]
@wgibbs-rs @tomoyanonymous as recent CoreAudio contributors, would you help to review?

File: src/host/coreaudio/macos/mod.rs [hunk 1]
```diff
@@ -930,26 +930,26 @@ fn set_sample_rate(
...
 pub struct Stream {
-    inner: Arc<Mutex<StreamInner>>,
+    inner: Rc<RefCell<StreamInner>>,
 }
 
 impl Stream {
     fn new(inner: StreamInner) -> Self {
         Self {
-            inner: Arc::new(Mutex::new(inner)),
+            inner: Rc::new(RefCell::new(inner)),
```
[the remark this answers is not in the document]
Comment (roderickvd) [reply src/host/coreaudio/macos/mod.rs:944]: This change is safe because:
 - The original `Arc<Mutex<StreamInner>>` provided no actual thread safety since `StreamInner` contains `AudioUnit` and `AudioObjectPropertyListener` which are not Send/Sync
 - CPAL's design is single-threaded for audio streams (verified through examples and usage patterns)
 - The clippy warning exists specifically to catch this pattern where `Arc` provides no benefit

The change makes the code semantically correct rather than misleadingly suggesting thread safety that didn't exist.

File: src/host/coreaudio/macos/mod.rs [hunk 2]
```diff
@@ -497,8 +497,8 @@ fn add_disconnect_listener<E>(
 where
     E: FnMut(StreamError) + Send + 'static,
 {
-    let stream_inner_weak = Arc::downgrade(&stream.inner);
-    let mut stream_inner = stream.inner.lock().unwrap();
+    let stream_inner_weak = Rc::downgrade(&stream.inner);
+    let mut stream_inner = stream.inner.borrow_mut();
```
[the remark this answers is not in the document]
Comment (roderickvd) [reply src/host/coreaudio/macos/mod.rs:497]: Fixed in [43e0a9d](https://github.com/RustAudio/cpal/pull/998/commits/43e0a9deaf5999a57ae0af57177ae62d44062673).

File: src/host/coreaudio/macos/mod.rs [hunk 3]
```diff
@@ -508,7 +508,7 @@ where
         },
         move || {
             if let Some(stream_inner_strong) = stream_inner_weak.upgrade() {
-                let mut stream_inner = stream_inner_strong.lock().unwrap();
+                let mut stream_inner = stream_inner_strong.borrow_mut();
                 let _ = stream_inner.pause();
```
[the remark this answers is not in the document]
Comment (roderickvd) [reply src/host/coreaudio/macos/mod.rs]: Fixed in [43e0a9d](https://github.com/RustAudio/cpal/pull/998/commits/43e0a9deaf5999a57ae0af57177ae62d44062673).

**Review by tomoyanonymous — COMMENTED** [review]
This is an additional suggetion but about the refactoring related to this, in `E` and `D` in `build_input/output_stream_raw` and `add_disconnect_listener`, `Send` is not necessary for them.

File: src/host/coreaudio/macos/mod.rs [hunk 4]
```diff
@@ -508,8 +504,15 @@ where
...
             if let Some(stream_inner_strong) = stream_inner_weak.upgrade() {
-                let mut stream_inner = stream_inner_strong.lock().unwrap();
-                let _ = stream_inner.pause();
+                match stream_inner_strong.try_borrow_mut() {
+                    Ok(mut stream_inner) => {
+                        let _ = stream_inner.pause();
+                    }
+                    Err(_) => {
+                        // Could not acquire mutable borrow; stream may already be in use.
+                        // Still notify about device disconnection even if we can't pause.
```
[the remark this answers is not in the document]
Comment (roderickvd) [reply src/host/coreaudio/macos/mod.rs]: Declined - Library shouldn't write to stderr; RefCell errors aren't informative enough to justify logging.

File: src/host/coreaudio/macos/mod.rs [hunk 5]
```diff
@@ -508,8 +504,15 @@ where
...
+                    }
+                }
                 (error_callback.lock().unwrap())(StreamError::DeviceNotAvailable);
```
[the remark this answers is not in the document]
Comment (roderickvd) [reply src/host/coreaudio/macos/mod.rs:518]: Incorrect - The `error_callback` must remain `Arc<Mutex<>>` because it needs to be `Send + 'static`. This is different from the stream data and the mixed approach is intentionally correct.

File: src/host/coreaudio/macos/mod.rs [hunk 6]
```diff
@@ -508,8 +504,15 @@ where
...
         move || {
             if let Some(stream_inner_strong) = stream_inner_weak.upgrade() {
-                let mut stream_inner = stream_inner_strong.lock().unwrap();
-                let _ = stream_inner.pause();
+                match stream_inner_strong.try_borrow_mut() {
+                    Ok(mut stream_inner) => {
+                        let _ = stream_inner.pause();
+                    }
+                    Err(_) => {
+                        // Could not acquire mutable borrow; stream may already be in use.
```
[the remark this answers is not in the document]
Comment (roderickvd) [reply src/host/coreaudio/macos/mod.rs]: Done in 8aa9a70.

**Comment by roderickvd** [comment]
> This is an additional suggetion but about the refactoring related to this, in `E` and `D` in `build_input/output_stream_raw` and `add_disconnect_listener`, `Send` is not necessary for them.

This is a very interesting point. While correct for CoreAudio, I believe that other backends (WASAPI, ALSA) *do* spawn threads where callbacks run, so the trait's `Send` bounds serve a real purpose there.

**Comment by tomoyanonymous** [comment]
Ah, I missed that the constraint came from `DeviceTrait`. Then it can be left them as they are. The other parts looks good to me.

**Comment by wgibbs-rs** [comment]
@roderickvd This looks great to me. I reviewed what you created for CoreAudio on my own device, and it worked first try. You fixed the mutability issues in StreamTrait, which is perfect, as well as everything else.

Previously, I tried working with multi-threading and it failed, and I was confused with the use of an Arc<Mutex<>>, but now that there's explicitly no-multithreading, I think that should be included in the documentation. Let me look and see if there's a place to put that in the docs.rs

**Comment by wgibbs-rs** [comment]
Never mind, I was mistaken on documentation. This is all internal, so documentation isn't necessary.

Apart from that, this looks great and runs on my MacBook Air M1.

**Comment by roderickvd** [comment]
Thanks for the quick reviews. I removed the changelog entry, as it the change isn't user-facing. Merging now.
